File: src/gtm_sourcing_agent/stages/screening.py
"""Stage 10: Recruiter Screening questions. See prompts/screening_questions.md."""

import json

from .. import llm_client, storage
from ..models import ScreeningQuestionSet


def run(role_id: str, candidate_id: str, *, storage_backend=storage) -> ScreeningQuestionSet:
    calibration = storage_backend.require_section(role_id, "calibration")
    candidates = storage_backend.require_section(role_id, "candidates")
    prioritizations = storage_backend.require_section(role_id, "prioritizations")
    if candidate_id not in candidates:
        raise ValueError(f"candidate '{candidate_id}' not found for role '{role_id}'")
    if candidate_id not in prioritizations:
        raise ValueError(
            f"candidate '{candidate_id}' has not been prioritized yet — run prioritize first"
        )

    prompt = llm_client.render_prompt(
        "screening_questions.md",
        candidate_json=json.dumps(candidates[candidate_id]),
        prioritization_json=json.dumps(prioritizations[candidate_id]),
        calibration_json=json.dumps(calibration),
    )
    result = llm_client.generate(prompt, ScreeningQuestionSet, stage="screening")
    result.candidate_id = candidate_id

    state = storage_backend.load_role(role_id)
    state.setdefault("screening", {})[candidate_id] = result.model_dump()
    storage_backend.save_role(role_id, state)
    return result

File: backend-node/src/stages/screening.ts
// Port of stages/screening.py -- Stage 10: Recruiter Screening questions.
import * as storage from "../db/storage.js";
import { StorageError } from "../db/storage.js";
import * as llmClient from "../llmClient.js";
import { ScreeningQuestionSet } from "../models.js";

export async function run(roleId: string, candidateId: string): Promise<ScreeningQuestionSet> {
  const calibration = await storage.requireSection(roleId, "calibration");
  const candidates = await storage.requireSection(roleId, "candidates");
  const prioritizations = await storage.requireSection(roleId, "prioritizations");
  if (!(candidateId in candidates)) {
    throw new StorageError(`candidate '${candidateId}' not found for role '${roleId}'`);
  }
  if (!(candidateId in prioritizations)) {
    throw new StorageError(`candidate '${candidateId}' has not been prioritized yet — run prioritize first`);
  }

  const prompt = llmClient.renderPrompt("screening_questions.md", {
    candidate_json: JSON.stringify(candidates[candidateId]),
    prioritization_json: JSON.stringify(prioritizations[candidateId]),
    calibration_json: JSON.stringify(calibration),
  });
  const result = await llmClient.generate(prompt, ScreeningQuestionSet, { stage: "screening" });
  result.candidate_id = candidateId;

  const state = await storage.loadRole(roleId);
  (state.screening ??= {})[candidateId] = result;
  await storage.saveRole(roleId, state);
  return result;
}
